binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java

```diff
@@ -123,11 +125,48 @@ Status checkAuthorization(MethodDescriptor<?, ?> method) {
           return authorization;
         }
       }
-      authorization = policy.checkAuthorizationForService(uid, serviceName);
+      try {
+        // TODO(10566): provide a synchronous version of "checkAuthorization" to avoid blocking the
+        // calling thread on the completion of the future.
+        authorization =
+            serverPolicyChecker.checkAuthorizationForServiceAsync(uid, serviceName).get();
+      } catch (ExecutionException e) {
+        // Do not cache this failure since it may be transient.
+        return Status.fromThrowable(e);
+      } catch (InterruptedException e) {
+        // Do not cache this failure since it may be transient.
+        Thread.currentThread().interrupt();
+        return Status.CANCELLED.withCause(e);
+      }
       if (useCache) {
         serviceAuthorization.putIfAbsent(serviceName, authorization);
       }
       return authorization;
     }
   }
+
+  /**
+   * Decides whether a given Android UID is authorized to access some resource.
+   *
+   * <p>This class provides the asynchronous version of {@link SecurityPolicy}, allowing
+   * implementations of authorization logic that involves slow or asynchronous calls without
+   * necessarily blocking the calling thread.
+   *
+   * @see SecurityPolicy
+   */
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/10566")
```

Comment:
This class is in `internal` package, so it doesn't need any experimental API annotation.